TBOX-162 movement import export

Tamiro2019 opened the pull request:
# ↪️ Pull Request

TBOX-162, First Pull Request, structure draft, intention is to gather initial comments for next iteration.

## ✔️ PR Todo

- [x] Add documentation
- [x] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [ ] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [ ] Pass all checks

Commit:
setting up import/export functionality stucture - based on Timo's code (see TBOX-152) - first draft

Commit:
adding docstrings and comments

Commit:
building data classes to fascilitate the construction of artifact lists to include/exclude when using import/export functions in movement.py

Commit:
adding asynchrnoous flag and wait function in operation.py, separating import into two api calls (new vs existing project) with required overwrite flag for existing project

Commit:
linting and format fixes

Tamiro2019 committed:
updating inits

Tamiro2019 committed:
updating doc strings and .rst files for project_artifacts

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+LOGGER = logging.getLogger(__name__)
+
+
+def export_project(
```
Adding `project` to the name might not be necessary as you are already within the `tbox.project` namespace.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+LOGGER = logging.getLogger(__name__)
+
+
+def export_project(
```
Adding `project` to the name might not be necessary as you are already within the `tbox.project` namespace.

Tamiro2019 replied:
Having export/import functions would be consistent. Unfortunately, there is a problem with using import(...) as a function : name is reserved by python. So I have kept export_artifacts, and import_artifacts for now; let me know if you have a better idea.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+    # check version compatibility for project movement
+    tamr_client = project.client
+    utils.version.enforce_after_or_equal(client=tamr_client, compare_version="2021.005.0")
+    logging.info("Tamr version has movement compatibility.")
```
This might be only needed for a debug log

Tamiro2019 replied:
removed logging.info("Tamr version has movement compatibility."), any errors would be raised by enforce_after_or_equal.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+
+    # make project export api request
+    body = {"artifactDirectory": artifact_directory_path, "excludeArtifacts": exclude_artifacts}
+    logging.info(f"Sending query to export project, {project.name}.")
```
Maybe "Starting to export project {project.name} of type {project.type}" is a better logging comment

Tamiro2019 replied:
Done; I've also changed import logging statements along the same lines.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+"""Tasks related to project movement as part of Tamr projects"""
```
I think the namespace used here is possibly confusing. To run the export of a mastering project you would have to call `tbox.project.mastering.movement.export_project`, whereas `tbox.project.mastering.export` would be make more sense for most users.

Tamiro2019 replied:
This makes sense. Changed __init__.py files per this request; please check.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+        operation for project export api call
+    """
+    # check version compatibility for project movement
+    tamr_client = project.client
```
`tamr_client` is not necessary here, you can use `project.client` everywhere instead.

Tamiro2019 replied:
Done (for exporting function). For the import function, it's a bit trickier to simplify in this way if we want the option of not specifying a Project object in the inputs, and automatically creating a new project.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+        utils.operation.enforce_success(operation)
+        logging.info("Export completed successfully.")
+
+    return operation
```
should the export function return the exported file location? There is currently no way for me to know the file location unless by looking inside the `operation` object that you return.

Tamiro2019 replied:
Currently the exported file directory must be specified in the input, but yeah, the full filepath is inside the operation object. 
During the planning discussion, I remember we said that only the operation object to be returned. 
I'm not sure if there is a good reason to only return the operation object, but it's worth discussing. 

For the moment added operation description to logging.info(f"...: {operation.description}.")

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+    # make project export api request
+    body = {"artifactDirectory": artifact_directory_path, "excludeArtifacts": exclude_artifacts}
+    logging.info(f"Sending query to export project, {project.name}.")
+    response = tamr_client.post(
```
how would you capture incorrect parameters or is it not possible to make a mistake? Potentially take inspiration from: https://github.com/Datatamer/tamr-toolbox/blob/812be755150c4198d1a52ac79c324b0d3146539b/tamr_toolbox/project/_common/transformations.py#L190

Tamiro2019 replied:
inspiration taken; for the moment it's a coarse capturer of incorrect parameters (response.status_code != 200); can be made finer if need be, but I reasoned that the details will be output in the error message anyways.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+    # make project import api request
+    body = {
+        "projectArtifact": project_artifact_path,
+        "newProjectName": new_project_name,
```
Is specifying `None` as a value to `NewProjectName` valid? Was this tested?

Tamiro2019 replied:
NewProjectName is required for importing into new projects, the api itself will through an error if this is None.
The json body now only contains NewProjectName if importing into new project.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,137 @@
+            raise KeyError("Unable to overwrite existing project; overwrite flag is off.")
+    else:
+        logging.info(f"Sending query to import new project, {new_project_name}.")
+        response = tamr_client.post("/v1/projects:import", json=body)
```
There a specific API to import a new project, what is the reasoning for using the same import API as for existing projects?

Tamiro2019 replied:
I'm not sure I understand the question. We are already using two separate APIs for import new project and for import into existing project.

pareantoine reviewed (commented):
@Tamiro2019, thanks for submitting your first PR to the Tamr Toolbox. Excellent first draft! Could you please add example scripts? I am not too sure how would the various dataclasses you have added would be used with the import and export APIs? Could you please look at how the transformation functions are added to the various projects by initialising them here for example (this is currently missing from your code): https://github.com/Datatamer/tamr-toolbox/blob/812be755150c4198d1a52ac79c324b0d3146539b/tamr_toolbox/project/mastering/__init__.py#L3

Tamiro2019 committed:
updating name package namespaces __init__'s

Tamiro2019 committed:
addressing comments on import and export functions

Tamiro2019 committed:
improving import logging

Tamiro2019 committed:
adding operation description to log

Tamiro2019 committed:
format fixes

Tamiro2019 committed:
improvements during testing

Tamiro2019 committed:
fixed broken api calls

Tamiro2019 committed:
adding snippets for movement import and export

Tamiro2019 committed:
adding import and export snippets

Tamiro2019 committed:
Merge branch 'develop' of https://github.com/Datatamer/tamr-toolbox into TBOX-162-movement-import-export
getting latest version of develop with test fixes

Tamiro2019 committed:
fixing long comments and unused imports

Tamiro2019 committed:
format fixes

Tamiro2019 committed:
fixing input empty list

Tamiro2019 committed:
commiting tests + minor changes to movement.py

Tamiro2019 committed:
completed tests

Tamiro2019 committed:
minor fixes

Tamiro2019 committed:
fixing config file reference

Tamiro2019 committed:
passed online and offline tests

Tamiro2019 committed:
fixing formatting

tiems90 commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,178 @@
+
+    # make project import api request
+    body = {
+        "projectArtifact": project_artifact_path,
```
Have you tested usage against projects that have spaces in their name? I think the artefact path here might require encoding with `%20` for spaces. 

One way to account for this is wrapping `project_artifact_path ` in `urllib.request.pathname2url()`.

Tamiro2019 replied:
Sounds good, thanks Timo; will test this soon.

tiems90 replied:
Just tested and confirmed it needs an encoding. 

```json
{
  "status": 400,
  "class": "javax.ws.rs.BadRequestException",
  "message": "Illegal character in path at index 29: /home/ubuntu/snapshot/Country Code Standardiser-1623667832395.zip"
...
}
```
where index 29 is the space in the path string. Only replacing spaces with %20 fixes the issue, but could still result in other issues for different characters.

This would affect all projects that have a space in their name.

along528 replied:
Discussed this with @pareantoine, @skalish. The decision was to include a workaround for this. Can be part of this PR or a separate parallel PR. 

At the same time will request this be fixed also in product. @tiems90 can you file a SUP ticket for this and link to this one?

The workaround should have a plan for backwards compatibility once the SUP is resolved

skalish replied:
As suggested by @tiems90, using `urllib.request.pathname2url()` is likely the safest way to do this.  When the necessary fix is incorporated into product, a new PR can be made to only do this when the software version is < the fix version.

Tamiro2019 replied:
Tried urllib.request.pathname2url(project_artifact_path) in import body, after exporting a project with spaces in name: '/home/ubuntu/tamr/projectExports/minimal schema mapping with spaces in name-1625100396077.zip'

attempting the import api urllib.request.pathname2url(project_artifact_path) gives: 
'Aborting import. The following validation errors are found:\nProject artifact /home/ubuntu/tamr/projectExports/minimal%2520schema%2520mapping%2520with%2520spaces%2520in%2520name-1625100396077.zip does not exist\n'

It seems like it does want the spaces in the path name. 
I'm not sure if I'm missing something, or if we did different things. 
Thoughts?

skalish replied:
`%25` is the encoding for `%`, so something strange does seem to be going on here.  Is it possible that typing the `%20` for spaces is necessary in swagger, but passing the string as part of a Python request does not require this?

skalish replied:
Essentially what I mean is that in order to get `%2520` from a space, something like `urllib.request.pathname2url` must be getting applied twice before the request makes it to the server.  @Tamiro2019 can you check if there's anyway this is happening?  Maybe the path with spaces already replaced is what is coming in as input to your function.

Tamiro2019 replied:
found the issue:
the tests were handling sequential export and import tasks, exports were already handling spaces, so imports where double handling them (hence %2520). Thanks for the observation Sam.
I've fixed this by wrapping with urllib.request.pathname2url() only when it detects spaces in the input string (not in general to avoid double wrapping effect, e.g. if the user already encoded the string before calling the import function).

pareantoine commented on examples/snippets/project/_common/export_artifacts.py:
```diff
@@ -0,0 +1,33 @@
+# Make list of categorization artifacts to exclude.
+# You can spell out the artifact code if known,
+# or list access via the CategorizationArtifacts dataclass
+exclude_list = [
```
very nice

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -32,42 +32,50 @@ def export_project(
     body = {"artifactDirectory": artifact_directory_path, "excludeArtifacts": exclude_artifacts}
-    logging.info(f"Sending query to export project, {project.name}.")
-    response = tamr_client.post(
+    logging.info(f"Starting to export {project.name} or type {project.type}.")
```
all your logging calls should be `LOGGER.` and not `logging.`

pareantoine replied:
typo? `or type` should be `of type`?

Tamiro2019 replied:
Fixed. Fixed.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -32,42 +32,50 @@ def export_project(
-    response.successful()
+    # Raise error if export api call not successful
+    if not response.ok:
+        error_message = f"Error exporting project artifacts: {response.json()['message']}"
```
don't limit the error message, supply the full response if possible

Tamiro2019 replied:
Removed ['message']

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
-        logging.info(f"Sending query to import new project, {new_project_name}.")
-        response = tamr_client.post("/v1/projects:import", json=body)
-    response.successful()
+        if new_project_name in [p.name for p in tamr_client.projects.stream()]:
```
You shouldn't need to `.stream()` `[p.name for p in tamr_client.projects]` is enough

Tamiro2019 replied:
removed .stream()

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
+        if new_project_name in [p.name for p in tamr_client.projects.stream()]:
+            error_message = f"New project name {new_project_name} already exists."
+            logging.error(error_message)
+            raise KeyError(error_message)
```
Is it a KeyError or a ValueError?

Tamiro2019 replied:
ValueError. Fixed.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
+            error_message = f"New project name {new_project_name} already exists."
+            logging.error(error_message)
+            raise KeyError(error_message)
+        if new_unified_dataset_name in [ds.name for ds in tamr_client.datasets.stream()]:
```
Same as above, no need to `.stream()` the `datasets` is already a collection

Tamiro2019 replied:
Fixed.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
+            # and artifacts will be partially migrated : creates issues on re-run.
+            error_message = f"New unified dataset name {new_unified_dataset_name} already exists."
+            logging.error(error_message)
+            raise KeyError(error_message)
```
same as previously, is it a KeyError or a ValueError

Tamiro2019 replied:
ValueError, fixed.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
         if overwrite_existing:
             logging.info(
-                f"importing existing project {target_project.name} "
+                f"Starting to import artifacts into existing project {target_project.name} "
```
I would log the path to the zipped artifacts in there too, just to be safe

Tamiro2019 replied:
Done.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
+            raise KeyError(error_message)
+        body["newProjectName"] = new_project_name
+        body["newUnifiedDatasetName"] = new_unified_dataset_name
+        logging.info(f"Starting to import artifacts into new project {new_project_name}.")
```
As previously, I would log the path to the zipped artifact in there too

Tamiro2019 replied:
Done.

pareantoine commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -93,36 +101,67 @@ def import_project(
+
+    # Raise exception if import was not successful
+    if not response.ok:
+        error_message = f"Error importing project artifacts: {response.json()['message']}"
```
log the full response

Tamiro2019 replied:
removed ['message']

pareantoine reviewed (changes requested):
@Tamiro2019, some really good changes in there and really simple examples. I still need to go through your unit tests but wanted to give you the opportunity to look at my comments first.

Tamiro2019 committed:
replacing logging's with LOGGER

Tamiro2019 committed:
addressing a few comments on response json printing, and re-running online tests without unnecessary recipe fix

Tamiro2019 commented:
> @Tamiro2019, some really good changes in there and really simple examples. I still need to go through your unit tests but wanted to give you the opportunity to look at my comments first.

> @pareantoine Thank you very much Antoine. I've gone through the comments, and made the corresponding fixes based on your suggestions. Re-ran the unit tests after removing an unnecessary recipe fix. Feel free to check the changes I've made, and to move on to reviewing the unit tests.

pareantoine commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+    artifact_path = str(description).split(" ")[-1]
+
+    assert artifact_path[0] == "/"
+    assert artifact_path[-4:] == ".zip"
```
Maybe use the following os function instead? https://docs.python.org/3/library/os.path.html#os.path.splitext

Tamiro2019 replied:
thanks for the recommendation

pareantoine commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+    assert artifact_path[-4:] == ".zip"
+
+    # import new project if it doesn't exist already
+    new_project_name = "new_categorization"
```
Should you add a timestamp to your new_project_name to make it trully unique?

Tamiro2019 replied:
I tried this and it works for live tests, but it becomes an issue when attempting to use the mock_api offline because every new test generates new timestamps, but the saved responses used by mock_api refer to past time-stamps, and the offline tests fail. 

For now I'm keeping the tests with definite pre-determined project names, without timestamp.

pareantoine commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+    assert artifact_path[-4:] == ".zip"
+
+    # import new project if it doesn't exist already
+    new_project_name = "new_categorization"
```
Should you add a timestamp to your new_project_name to make it trully unique?

Tamiro2019 replied:
same as above, I found it best to create unique pre-determined names without timestamps to get successful offline tests with the mock_api().

pareantoine commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+    # import new project if it doesn't exist already
+    new_project_name = "new_categorization"
+    new_unified_dataset_name = new_project_name + "_ud"
+    if new_project_name not in [p.name for p in client.projects.stream()]:
```
you don't need to add `.stream()`

Tamiro2019 replied:
right, thanks!

pareantoine reviewed (commented):
Unit tests look good @Tamiro2019.
Seems that you're going through every possibility in the tests which is good.

Just a few comments here and there, but otherwise LGTM.

along528 commented on examples/snippets/project/_common/export_artifacts.py:
```diff
@@ -0,0 +1,33 @@
+"""Snippet for exporting project artifacts from a Tamr project"""
+import tamr_toolbox as tbox
+from tamr_toolbox.project import categorization
+from tamr_toolbox.models.project_artifacts import CategorizationArtifacts as catfacts
```
```suggestion
from tamr_toolbox.models.project_artifacts import CategorizationArtifacts
```

Prefer clarity over brevity

Tamiro2019 replied:
preference acknowledged (and change accepted)

along528 commented on examples/snippets/project/_common/import_artifacts.py:
```diff
@@ -0,0 +1,36 @@
+import tamr_toolbox as tbox
+from tamr_toolbox.project import categorization
+
+# from tamr_toolbox.models.project_artifacts import CategorizationArtifacts as catfacts
```
Can this comment be removed?

Tamiro2019 replied:
yes, done

along528 commented on tamr_toolbox/models/project_artifacts.py:
```diff
@@ -0,0 +1,162 @@
+
+
+@dataclass()
+class Artifacts:
```
Artifacts should only be defined in one place. If this only included for convenience I suggest removing this dataclass. If you believe it is necessary then it should be constructed from the union of all other project specific dataclasses

Tamiro2019 replied:
I don't think this is needed, users should be able to find artifacts based for specific project types.
Removed extra class Artifacts.

along528 commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,182 @@
+        new_project_name: new project name
+        new_unified_dataset_name: new unified dataset name
+        exclude_artifacts: list of artifacts to exclude in import
+        include_additive_artifacts: list of artifacts to import only additively
```
Only certain artifacts support being imported additively. See
https://docs.tamr.com/new/docs/project-movement#working-with-project-artifacts

Ideally the API throws an error when attempting to import additively for an artifact that doesn't support it. If so, include a unit test for this. If not then may need to add some checks in the toolbox

Tamiro2019 replied:
The API throws an error when trying to additively include an artifact which is not supported.
Added to test to catch such error when it occurs.

along528 commented on tamr_toolbox/project/_common/movement.py:
```diff
@@ -0,0 +1,182 @@
+        operation for project import api call
+    """
+    # initializing empty lists if any
+    if exclude_artifacts is None:
+        exclude_artifacts = []
+    if include_additive_artifacts is None:
+        include_additive_artifacts = []
+    if include_destructive_artifacts is None:
+        include_destructive_artifacts = []
```
If empty does this result in the default behavior listed here?
https://docs.tamr.com/new/docs/project-movement#working-with-project-artifacts 

Or are they excluded? It should be the default behavior

Tamiro2019 replied:
If empty, the default behavior occurs.

along528 commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+        )
+
+    # fail if not present
+    with pytest.raises(ValueError):
```
Shouldn't this always fail if not present? What is the current behavior?

Tamiro2019 replied:
My understanding is that, by default, artifacts can be imported into a project even if those artifacts do not exist in the project. When the failIfNotPresent flag does is it makes the import fail if an artifact exists in the zip which does not exist in the target project.

along528 commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+            overwrite_existing=True,
+        )
+    # Expected ValueError but got RuntimeError instead
+    with pytest.raises(RuntimeError):
```
What is this testing? Appears to be the same as call as the previous one

Tamiro2019 replied:
subtle difference: include_destructive_artifacts vs include_additive_artifacts. Both are tested incorrect artifact names in list.

along528 commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+
+
+@mock_api()
+def test_import_new():
```
Need to add at least one test that covers succesfull import when specifying specific artifacts to import. Both for import new and import existing

Tamiro2019 replied:
Included in tests for import new and existing

along528 commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,437 @@
+import pytest
+from tamr_toolbox import utils, workflow
+from tests._common import get_toolbox_root_dir
+from tamr_toolbox.project import categorization
+from tamr_toolbox.models.project_artifacts import CategorizationArtifacts as catfacts
```
Current tests only cover categorization. The other project types need to be tested as well. You might be able to use your existing unit tests by parameterizing the tests by project type using the `pytest.mark.parametrize` decorator. Example here:
https://github.com/Datatamer/tamr-toolbox/blob/develop/tests/data_io/test_csv.py#L75

Tamiro2019 replied:
Thanks for the suggestion, very useful decorator!
Added tests for mastering, schema mapping, and golden records.

along528 replied:
Fantastic!

along528 reviewed (changes requested):
The PR is looking very good! Have a few comments/questions around the behavior. Biggest request is that there is still a need to extend the test coverage to cover not just categorization project types but all project types

along528 commented on tamr_toolbox/models/project_artifacts.py:
```diff
@@ -0,0 +1,162 @@
+
+
+@dataclass()
+class SchemaMappingArtifacts:
```
Notice that you need to specify the type as str for each field. For example:
`UNIFIED_ATTRIBUTES: str = "UNIFIED_ATTRIBUTES"`

If you don't then you can't loop over the fields. See this for example
```
In [11]: from dataclasses import dataclass, fields                                                                                                                                                            

In [12]: @dataclass 
    ...: class A: 
    ...:   x: int = 1 
    ...:   y: int = 2 
    ...:                                                                                                                                                                                                      

In [13]: [field.name for field in fields(A)]                                                                                                                                                                  
Out[13]: ['x', 'y']

In [14]: @dataclass 
    ...: class A: 
    ...:   x = 1 
    ...:   y = 2                                                                                                                                                                                              

In [15]: [field.name for field in fields(A)]                                                                                                                                                                  
Out[15]: []
```

Tamiro2019 replied:
Fixed

Tamiro2019 committed:
addressing a few comments on export with spaces in names + adding testing for export with spaces in names, to test errors with unsupported additive artifacts, and to parameterize import tests for new projects of different types.

Tamiro2019 committed:
adding parametrization to test across multiple project types and explicit artifact choices for import

Tamiro2019 committed:
fixing linting and formatting issues

tiems90 commented:
It was required in my python requests against 2021.10
________________________________
From: skalish ***@***.***>
Sent: Thursday, July 1, 2021 10:03:50 PM
To: Datatamer/tamr-toolbox ***@***.***>
Cc: Timo ***@***.***>; Mention ***@***.***>
Subject: Re: [Datatamer/tamr-toolbox] TBOX-162 movement import export (#36)


@skalish commented on this pull request.

________________________________

In tamr_toolbox/project/_common/movement.py<https://github.com/Datatamer/tamr-toolbox/pull/36#discussion_r662590757>:

> +        operation for project import api call
+    """
+    # initializing empty lists if any
+    if exclude_artifacts is None:
+        exclude_artifacts = []
+    if include_additive_artifacts is None:
+        include_additive_artifacts = []
+    if include_destructive_artifacts is None:
+        include_destructive_artifacts = []
+
+    # check version compatibility for project movement
+    utils.version.enforce_after_or_equal(client=tamr_client, compare_version="2021.005.0")
+
+    # make project import api request
+    body = {
+        "projectArtifact": project_artifact_path,


%25 is the encoding for %, so something strange does seem to be going on here. Is it possible that typing the %20 for spaces is necessary in swagger, but passing the string as part of a Python request does not require this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Datatamer/tamr-toolbox/pull/36#discussion_r662590757>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEGGGVWVNHSIBX45KZE2JA3TVTJ3NANCNFSM44ITP63A>.

Tamiro2019 committed:
fixing tests for handling golden records, new project timestamping, and name in spaces

Tamiro2019 committed:
adding ndjsons for parametric tests (import/export over multiple project types)

Tamiro2019 committed:
removing outdated ndjson files

skalish commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,539 @@
+            ],
+        ),
+        (
+            CONFIG["projects"]["minimal_mastering"],
```
You can make the test parameters as passed here simpler to shorten the test filenames.  For instance, instead of
```python
        (
            CONFIG["projects"]["minimal_mastering"],
            "new_mastering_2",
            [MasteringArtifacts.RECORD_PAIR_VERIFIED_LABELS],
            [MasteringArtifacts.CLUSTERING_MODEL, MasteringArtifacts.PUBLISHED_CLUSTERS],
        )  
```
you could have
```python
        (
            "minimal_mastering",
            [MasteringArtifacts.RECORD_PAIR_VERIFIED_LABELS],
            [MasteringArtifacts.CLUSTERING_MODEL, MasteringArtifacts.PUBLISHED_CLUSTERS],
        )  
```
and then in your test have `CONFIG["projects"][project_to_export]` in place of `project_to_export` as it is now.  You can have a dictionary that maps, e.g. `"minimal_mastering"` to `"new_mastering_2"` and use that to look up the correct `new_project_name` value in each case.

Tamiro2019 replied:
Your solution worked like a charm:
adding
```
      - name: Support longpaths
        run: git config --system core.longpaths true
```
at the beginning of the
```
  Test-Windows:
    ...
    steps:
```
section (around line 70) of .github/workflows/ci.yml

Tamiro2019 committed:
trying to change configuration ci.yml to allow for long paths

along528 commented on tests/project/_common/test_movement.py:
```diff
@@ -0,0 +1,539 @@
+    existing_project = client.projects.by_name(existing_project_name)
+
+    # test import into existing project
+    op = categorization.import_artifacts(
```
The `categorization` module is being used here for export/import of all project types. This exposes an unnecessary constraint on how this is being called i.e. the api is the same regardless of which module you import. The preferred behavior I believe would be not to require calling the correct project type (since it's not actually required to do this). But to make the mechanism for calling this common to all projects. For example: `tbox.projects.export_artifacts`

@pareantoine @skalish @tiems90 what do you think?

along528 replied:
Resolved with https://github.com/Datatamer/tamr-toolbox/pull/60

along528 committed:
TBOX-162 Use common namespace for Project Movement (#60)

* Change namespace for movement calls
project.<project type>.import_artifacts becomes project.import_artifacts
project.<project type>.export_artifacts becomes project.export_artifacts

* Add to documentation and examples

* Linting

along528 reviewed (approved):
LGTM! Great work Tamiro

Before merging should wait for approval from Antoine

pareantoine reviewed (approved):
LGTM. Amazing first addition to the toolbox @Tamiro2019, this will be used by many.

tiems90 reviewed (approved):
Looks great, @Tamiro2019! Can't wait to put it to use!

Tamiro2019 commented:
Thanks everyone for reviewing this PR and all your super helpful comments :)
